src/extension/geofence/src/main/java/org/geoserver/geofence/GeofenceAccessManager.java

```diff
@@ -269,108 +279,189 @@ public StyleAccessLimits getAccessLimits(Authentication user, StyleInfo style) {
-            ruleFilter.setSourceAddress(RuleFilter.SpecialFilterType.DEFAULT);
+    private String getUserNameFromAuth(Authentication authentication) {
+        String username = authentication != null ? authentication.getName() : null;
+        if (username != null && username.isEmpty()) {
+            username = null;
         }
+        return username;
+    }
 
-        LOGGER.log(Level.FINE, "ResourceInfo filter: {0}", ruleFilter);
-
-        AccessInfo rule = rules.getAccessInfo(ruleFilter);
-
-        if (rule == null) {
-            rule = AccessInfo.DENY_ALL;
+    private Pair<Geometry, AccessInfo> handleContainers(
+            List<LayerGroupInfo> groups, Authentication user) {
+        Geometry allowedArea = null;
+        AccessInfo lessRestrictiveRule = null;
+        for (LayerGroupInfo lgi : groups) {
+            WorkspaceInfo ws = lgi.getWorkspace();
+            String workspace = ws != null ? ws.getName() : null;
+            String layer = lgi.getName();
+            RuleFilter filter = buildRuleFilter(workspace, layer, user);
+            AccessInfo accessInfo = rules.getAccessInfo(filter);
+            lessRestrictiveRule = getLessRestrictiveRule(lessRestrictiveRule, accessInfo);
+            allowedArea =
+                    getReprojectArea(accessInfo, lgi.getBounds().getCoordinateReferenceSystem());
         }
+        return new ImmutablePair<>(allowedArea, lessRestrictiveRule);
+    }
 
-        DataAccessLimits limits = buildAccessLimits(resource, rule);
-        LOGGER.log(
-                Level.FINE,
-                "Returning {0} for layer {1} and user {2}",
-                new Object[] {limits, resource.prefixedName(), username});
+    // get the AccessInfo from the layer's containers returning a tuple with
+    // the Geometry of the allowed area if found and the less restrictive accessInfo
+    // among the ones of associated layerGroups
+    private Pair<Geometry, AccessInfo> getContainerRuleForLayerDirectAccess(
+            Object resource, Authentication user) {
+        MutablePair<Geometry, AccessInfo> result = new MutablePair<>(null, null);
+        Collection<LayerGroupContainmentCache.LayerGroupSummary> summaries;
+        if (resource instanceof ResourceInfo)
+            summaries = groupsCache.getContainerGroupsFor((ResourceInfo) resource);
+        else if (resource instanceof LayerInfo)
+            summaries = groupsCache.getContainerGroupsFor(((LayerInfo) resource).getResource());
+        else summaries = groupsCache.getContainerGroupsFor((LayerGroupInfo) resource);
+
+        for (LayerGroupContainmentCache.LayerGroupSummary gs : summaries) {
+            LayerGroupInfo.Mode mode = gs.getMode();
+            if (mode.equals(LayerGroupInfo.Mode.OPAQUE_CONTAINER)) {
+                // opaque mode deny access for the layer
+                AccessInfo newInfo = AccessInfo.DENY_ALL;
+                result.setValue(getLessRestrictiveRule(result.getRight(), newInfo));
```

Comment:
Confusing usage of Pair API. Decide if you want to use left/right or key/value and use them consistently.

Reply:
fixed